src/factory/src/api.rs

```diff
@@ -2,126 +2,135 @@
+        self.state.replace(State::new(controller, ledger_principal));
     }
 
-    set_token_bytecode_impl(bytecode).await;
-}
-
-/// Creates a new token.
-///
-/// Creating a token canister with the factory requires one of the following:
-/// * the call must be made through a cycles wallet with enough cycles to cover the canister
-///   expenses. The amount of provided cycles must be greater than `10^12`. Most of the cycles
-///   will be added to the newly created canister balance, while some will be consumed by the
-///   factory
-/// * the caller must transfer some amount of ICP to their subaccount into the ICP ledger factory account.
-///   The subaccount id can be calculated like this:
-///
-/// ```ignore
-/// let mut subaccount = [0u8; 32];
-/// let principal_id = caller_id.as_slice();
-/// subaccount[0] = principal_id.len().try_into().unwrap();
-/// subaccount[1..1 + principal_id.len()].copy_from_slice(principal_id);
-/// ```
-///
-/// The amount of provided ICP must be greater than the `icp_fee` factory property. This value
-/// can be obtained by the `get_icp_fee` query method. The ICP fees are transferred to the
-/// principal designated by the factory controller. The canister is then created with some
-/// minimum amount of cycles.
-///
-/// If the provided ICP amount is greater than required by the factory, extra ICP will not be
-/// consumed and can be used to create more canisters, or can be reclaimed by calling `refund_icp`
-/// method.
-#[update(name = "create_token")]
-#[candid_method(update, rename = "create_token")]
-pub async fn create_token(
-    info: Metadata,
-    owner: Option<Principal>,
-) -> Result<Principal, TokenFactoryError> {
-    if info.name.is_empty() {
-        return Err(TokenFactoryError::InvalidConfiguration(
-            "name",
-            "cannot be `None`",
-        ));
+    /// Returns the token, or None if it does not exist.
+    #[query]
+    async fn get_token(&self, name: String) -> Option<Principal> {
+        self.state.borrow().factory.get(&name)
     }
 
-    if info.symbol.is_empty() {
-        return Err(TokenFactoryError::InvalidConfiguration(
-            "symbol",
-            "cannot be `None`",
-        ));
+    pub async fn set_token_bytecode_impl(&self, bytecode: Vec<u8>) {
+        self.state.borrow_mut().token_wasm.replace(bytecode);
     }
 
-    let state = State::get();
-    let key = info.name.clone();
-
-    if state.borrow().factory.get(&key).is_some() {
-        return Err(TokenFactoryError::AlreadyExists);
+    #[update]
+    async fn set_token_bytecode(&self, bytecode: Vec<u8>) {
+        if self.state.borrow().controller() != ic_kit::ic::caller() {
+            #[cfg(target_family = "wasm")]
+            {
+                ic_cdk::api::call::reject("not authorized");
+                return;
+            }
+            #[cfg(not(target_family = "wasm"))]
+            panic!("not authorized");
```

Comment:
Since you already use `ic-kit`, use `ic_kit::ic::trap` instead. It does the same thing.